addon/components/frost-popover.js

```diff
@@ -1,96 +1,328 @@
 import Ember from 'ember'
 import layout from '../templates/components/frost-popover'
 import $ from 'jquery'
-export default Ember.Component.extend({
+import PropTypeMixin, {PropTypes} from 'ember-prop-types'
+import {checkLeft, checkRight, checkTop, checkBottom} from './util'
+
+const arrowMargin = 5
+const maxPlacementRetries = 5
+
+export default Ember.Component.extend(PropTypeMixin, {
   layout,
-  visible: false,
-  event: 'click',
-  closest: false,
-  offset: 10,
-  position: 'bottom',
-  index: 0,
-  excludePadding: false,
-  classNameBindings: ['position', 'visible:visible:invisible'],
+  classNameBindings: ['visible:visible:invisible', 'autoPosition'],
   classNames: ['tooltip-frost-popover'],
-  didRender () {
+  propTypes: {
+    closest: PropTypes.bool,
+    event: PropTypes.string,
+    excludePadding: PropTypes.bool,
+    index: PropTypes.number,
+    offset: PropTypes.number,
+    position: PropTypes.string,
+    resize: PropTypes.bool,
+    viewport: PropTypes.oneOfType([
+      PropTypes.string,
+      PropTypes.object,
+      PropTypes.func
+    ])
+  },
+
+  getDefaultProps () {
+    return {
+      closest: false,
+      event: 'click',
+      excludePadding: false,
+      index: 0,
+      offset: 10,
+      position: 'bottom',
+      resize: true,
+      viewport: 'body',
+      visible: false,
+      autoPosition: 'bottom'
+    }
+  },
+
+  didInsertElement () {
     Ember.run.next(() => {
-      const context = this
-      if (this.get('closest')) {
-        this.$().closest(this.get('target')).on(this.get('event'), function () {
-          context.send('togglePopover')
-        })
+      const target = this.getTarget()
+      const event = this.get('event')
+
+      this._eventHandler = (event) => {
+        const popover = this.get('element')
+        if ($(event.target).closest(popover).length === 0) {
+          this.send('togglePopover')
+        }
+      }
+
+      $(target).on(event, this._eventHandler)
```

Comment:
That makes a lot of sense. We currently can't use this as it is to replace `frost-select`'s dropdown because consumers can't control the visibility state. On the other hand, tooltips should be dead simple to use so I picture it as more of a transient. There are some features of this component that makes it compelling to be used as a generic dropdown (e.g. viewport containment and auto placement) but consumers who just want a simple replacement to the `title` attribute to show tooltips will have to incur that complexity. Perhaps there's an opportunity for another component `ember-frost-dropdown`. `ember-frost-popover` can extend from that to handle the state for the consumers for the simple transient case.